### DISCUSSION TITLE
ESP32 examples Mongoose V7.9

### DISCUSSION THREAD

**Discussion by alwin4711** [cesanta/mongoose#2195] [Ask questions, get answers] [answered]
Hello,
i want to know if any one has compiled the esp32 examples from mangoose version 7.9 ?
I work with VSCODE and platformio.
example: /esp32/device_dashboard
without Makefile (i don't use docker) just compile with the esp-idf.
i use -DMG_ARCH = ESP_PLATFORM and -DMG_ARCH = MG_ARCH_ESP32
as comiler flags
some errors for  examples:

main/mongoose.h:357:10: fatal error: poll.h: No such file or directory

main/net.c:204:7: error: 'struct mg_connection' has no member named 'label'
  204 |      c->label[0] = 'W';  // Mark ourselves as a event listener

main/mongoose.h:811:19: error: incompatible type for argument 2 of 'mg_mqtt_sub'
  811 | #define mg_str(s) mg_str_s(s)
      |                   ^~~~~~~~~~~
      |                   |
      |                   struct mg_str
main/net.c:107:25: note: in expansion of macro 'mg_str'
  107 |     mg_mqtt_sub(s_mqtt, mg_str(s_config.sub), 2);

maybe this examples need an update ?

Thnks alot for any help.
Best regards
Alain

**Comment by scaprile** [comment, marked as answer]
> Hello, i want to know if any one has compiled the esp32 examples from mangoose version 7.9 ?

Yes, Github does that for us every time we introduce a new line of code, all our examples build.

> I work with VSCODE and platformio.

VSCode is just an editor, but PlatformIO is probably a different architecture and we don't have working examples for that, we don't even know if Mongoose will work on that.

> example: /esp32/device_dashboard without Makefile (i don't use docker) just compile with the esp-idf. i use -DMG_ARCH = ESP_PLATFORM and -DMG_ARCH = MG_ARCH_ESP32 as comiler flags some errors for examples:

Where did you get that from ? Your first choice is not a valid Mongoose platform but an IDF defined symbol, and the second one should not be needed, as the platform is automatically detected based on the former symbol being defined by the IDF.
https://github.com/cesanta/mongoose/blob/9c0fe8999d988bcc9485c5402868291017ec21eb/src/arch.h#L27-L28

Please read the docs https://mongoose.ws/documentation/#build-options

> main/mongoose.h:357:10: fatal error: poll.h: No such file or directory

Line 357 is different both in 7.9 stable and git head. What is your release ?
Anyway, that line number belongs to MG_ARCH_UNIX and you are not compiling for a Un*x system so that is your problem.

> main/net.c:204:7: error: 'struct mg_connection' has no member named 'label' 204 | c->label[0] = 'W'; // Mark ourselves as a event listener

We moved `label` to `data` long ago, if you read the docs you'll see there is really no member named 'label' as the error says. https://mongoose.ws/documentation/#struct-mg_connection
You seem to be trying to compile an example from an older release with Mongoose sources from a newer release. Your line 204 does not match neither 7.9 nor git head, again.

Please follow the tutorials https://mongoose.ws/documentation/tutorials/esp32/device-dashboard/

> maybe this examples need an update ?

Not from us, please get your code from the repo as the tutorials indicate

If you have a specific issue, open an issue, and fill the template. However, we are not currently supporting PlatformIO, so we'll need the IDF version you are using.
If you want to port to a different architecture, you are on your own, we may help if you ask specific questions.

**Reply by alwin4711** [reply]
Hi Sergio,
thank you for your reply.

ok, i got a older archive file for mongoose from the espressif webpage ;-)
now i take the right one an this code looks much better.
Now i got the problem of include <poll.h> but this is ok. ;-) 

I have to read much more of the excelent documentation.
Anything from "Build tools" and "Tutorial and examples"

The only thing i dont like is to use docker for the esp exaples ;-) 
But that can i change when i will use the espressif-ide for the ESP32 part.

Ciao,
Alain

**Reply by scaprile** [reply]
> The only thing i dont like is to use docker for the esp exaples ;-) But that can i change when i will use the espressif-ide for the ESP32 part.

Docker is there to keep a stable version of the IDF, our CI/CD picks that up and tests our code builds OK.
You are not obliged to use Docker, if you know how to use the IDF you can just use the IDF, the Makefile just calls the IDF inside a Docker image.
https://github.com/cesanta/mongoose/blob/9c0fe8999d988bcc9485c5402868291017ec21eb/examples/esp32/device-dashboard/Makefile#L3-L13
All that this makefile does is to run `idf.py build` without having to install the IDF first. The project is a plain IDF project, you can build with anything that is able to properly build IDF projects (and sets the proper expected symbols)

**Reply by alwin4711** [reply]
Hi Sergio,
a short response about my problems.....
READING the DOCs is the matchmaker ;-)) 
(special the part with the symlinks)
Everthing is compiled and work is in progress...
Thank you very much for your help.
Best regards
A,lain

**Reply by scaprile** [reply]
Thx 4 letting us know
